### ISSUE TITLE
PortPerBrokerClusterNetworkAddressConfigProvider should start with count rather than nodeId

### ISSUE THREAD

**Issue by callaertanthony** [kroxylicious/kroxylicious#902]
**Describe the bug**
Currently, it's hard to maintain

**To Reproduce**
1. Configure kroxylicious like this ...
2. Run command '....'
3. See error

**Expected behavior**

**YAML files and logs**

Attach or copy paste your kroxylicious configuration YAML (with sensitive details
like broker address redacted).
Attach or copy and paste relevant logs.

**Additional context**

**Comment by robobario** [comment]
It's a tricky problem. 

A reason it's like this is the client must be able to address every broker. By having a deterministic mapping like this we can have multiple instances of the proxy and rely on port X mapping to broker Y on every proxy instance without them sharing state. The clients can rely on `proxyHost:${portX}` always mapping to broker Y, which will match what we've said in previous Metadata responses. This is important when we do things like restart the proxy, the clients will potentially try to reconnect on those ports expecting it to be the same broker, rather than using the bootstrap address, here's an issue with some discussion about that [problem](https://github.com/kroxylicious/kroxylicious/issues/416).

To do this, on startup we open a port for every possible `nodeId` in our configurable `numberOfBrokerPorts`. So if `numberOfBrokerPorts` is 6 the proxy opens 6 ports immediately (with the assumption the ids will be 1-6), regardless of what brokers really exist. If a client connects to any of those ports we will proxy to the upstream bootstrap to obtain metadata, intercept it and set up the mappings, then respond to the client with updated metadata and close that connection. Then on the subsequent connection attempt the port should be mapped to the correct upstream broker (or the port will be closed IIRC if that broker is no longer active).

We could add a configurable `startNodeId` to at least handle the [1000, 1001, 1002, ...]  case.

**Comment by robobario** [comment]
now I'm wondering if it's really okay to close a port once we discover there's no active broker :grinning: we might help one client recover, but then the next client can't reconnect on any of the old ports.

**Comment by callaertanthony** [comment]
In my opinion the most common case is a rolling update, and Kroxy would be able to handle and fallback the common error to the client. Then, the client will just fetch metadata again, and call to new assigned partition, etc.

Not sure when client uses an old port assigned to a new node is a real problem. Considering the client uses not updated metadata. Kroxy will remap upstream from the requested port. If upstream validates the request, Kroxy will answers to client at is. If not, like a non leader partition answer, the error will fallback to client, and it should re-ask for metadata no ?

From kafka [documentation](https://kafka.apache.org/protocol#protocol_partitioning) :

> The client does not need to keep polling to see if the cluster has changed; it can fetch metadata once when it is instantiated cache that metadata until it receives an error indicating that the metadata is out of date. This error can come in two forms: (1) a socket error indicating the client cannot communicate with a particular broker, (2) an error code in the response to a request indicating that this broker no longer hosts the partition for which data was requested.
	1. Cycle through a list of "bootstrap" Kafka URLs until we find one we can connect to. Fetch cluster metadata.
	2. Process fetch or produce requests, directing them to the appropriate broker based on the topic/partitions they send to or fetch from.
	3. If we get an appropriate error, refresh the metadata and try again.
	
I will try to expose the problem to address with a sequence graph to help the discussion going on.

**Comment by robobario** [comment]
Good point. The above issue we saw was that if all the broker ports on the proxy are removed, never to return, then the clients cannot recover. But if we mapped them as in your issue's example then they should reconnect, request metadata and do something sensible with the updated mappings. Or hit non-leader exceptions.

Are you imagining it being configured in, like the proxy would have a configured fixed mapping? Or that you would keep it dynamic (use intercepted MetadataResponse to map the topology) and rely on the proxies being eventually consistent? They would be driven towards the same state by clients hitting non leader partition and requesting metadata, eventually being consistent with the upstream?

I guess we would need code to drop established connections in the proxy when mappings change. Like if proxy port 1234 is mapped to upstream node 2 (node 1 is inactive), upstream node 1 comes up and is mapped to port 1234, proxy should disconnect channels on 1234

**Comment by robobario** [comment]
Maybe a non-issue but could it happen that a consumer un-knowningly starts fetching from a follower increasing the latency a little?

**Comment by callaertanthony** [comment]
> Good point. The above issue we saw was that if all the broker ports on the proxy are removed, never to return, then the clients cannot recover. But if we mapped them as in your issue's example then they should reconnect, request metadata and do something sensible with the updated mappings. Or hit non-leader exceptions.

This problem sounds more like a DRP. When this happen, it feels really hard in a real world to don't have manual action (offset management, consumer restart, etc.). --> if we consider we must map new nodes.
So it's hard for the proxy to fully handle this use case.

> Are you imagining it being configured in, like the proxy would have a configured fixed mapping? Or that you would keep it dynamic (use intercepted MetadataResponse to map the topology) and rely on the proxies being eventually consistent? They would be driven towards the same state by clients hitting non leader partition and requesting metadata, eventually being consistent with the upstream?

If I'm not wrong, at this time proxy remap upstream port on each request,  so client could handle new nodes arriving on cluster, it can handle node eviction too.

I guess that the alone problem existing today is the port mapping strategy. By summing the node id to opened port, proxy asks to operator opening an infinite port range, because itself will never stop to increment node port (++nodeId).
With a mathematical point of view as for exemple, if the mapping strategy is a modulo rather than a sum, this problem should not exists, because proxy will be limited on port count by modulo.

Certainly not the best solution, but something like this works to expose what I mean :
```java
private int evaluateNodeId(VirtualCluster virtualCluster, int nodeId) {
    var portCount = virtualCluster.getClusterNetworkAddressConfigProvider().getExclusivePorts().size() - 1;
    if (nodeId > portCount) {
        return nodeId % portCount;
    }
    return nodeId;
}
```

**Comment by robobario** [comment]
Modulo sounds like something we could implement. It would have the caveat that if you set the number of broker ports to N, then you'd want `maxNodeId - minNodeId < N` at any time for there to be a deterministic mapping. For example if you've configured the proxy with `numberOfBrokerPorts = 6` the modulo strategy would be broken if there were upstream nodes with ids `1, 7`. We could potentially make the proxy detect this case when we intercept the upstream Metadata and take some action.

Maybe we could add it as an optional behaviour like:

```
clusterNetworkAddressConfigProvider:
  type: PortPerBrokerClusterNetworkAddressConfigProvider
  config:
    bootstrapAddress: mycluster.kafka.com:9192 
    numberOfBrokerPorts: 6
    portMappingStrategy: moduloNodeId|nodeId
```

**Comment by robobario** [comment]
Another thing on the horizon is admin clients being able to communicate with KRaft Controllers, See [KIP-919](https://cwiki.apache.org/confluence/display/KAFKA/KIP-919%3A+Allow+AdminClient+to+Talk+Directly+with+the+KRaft+Controller+Quorum+and+add+Controller+Registration). If the controllers and brokers are separate nodes with distinct ranges of node.id that would be another thing to consider.

**Comment by k-wall** [comment]
I think to solve this problem, `portperbroker` needs knowledge of the node ids that the cluster is actually using.

Kroxylicious could **discover** the information from the cluster.  That would be nice as it keeps the amount of configuration required to stand up the proxy the down to a minimum.   However, Kroxylicious would need credentials/permission to connect to the cluster in its own right.  We'd also need to think about behaviour if the cluster changed shape at runtime. 

The alternative is to configure Kroxylicious with **a priori** knowledge of the node ids that the cluster is using.  It could be defined in terms of node id ranges which would allow room for cluster to grow without having to reconfigure Kroxylicious. This would help mitigate some of the inconvenience of the extra configuration.

It could look something like this:

```
        clusterNetworkAddressConfigProvider:
          type: RangeAwarePortPerBroker
          config:
            bootstrapAddress: localhost:9292
            brokerAddressPattern: kroxylicious-service
            brokerStartPort: 9293 # defaults to bootstrap port +1
            nodeIdRanges:
            - name: controllers    
               range: [0,10]
            - name: brokers
               range: [100,110)
            - name: morebrokers
               range: [200,210)
```

It would be an error to define a ranges that intersect.   The range name is an arbitrary label. It must be unique.  It will be used to produce error messages in the event of a range definition error.  The range name has no effect on processing.

The range itself would be defined using interval notation. 
The order in which the ranges are defined in the nodeIdRanges list is not significant.

When the proxy needs to allocate a port for a broker with node id _n_, it will compute:

```
offset = indexOf(nodeIdRange, n)
broker-port = brokerStartPort + offset
```

This will mean that port numbers will be allocated consecutively even if there are gaps in the node id ranges used by the cluster.

I think this would resolve the issue that @callaertanthony reported above.  It should also help @luozhenyu use-case discussed in https://github.com/kroxylicious/kroxylicious/pull/1175#issuecomment-2097578792.   I also think this should work for Strimzi users taking advantage of [node id management](https://strimzi.io/blog/2023/08/23/kafka-node-pools-node-id-management/).
WDYAT?

**Comment by robobario** [comment]
To be explicit when you say `offset = indexOf(nodeIdRange, n)`, are you flattening them into an ordered array? So:

```
            nodeIdRanges:
            - name: brokers
               range: [4,6)
            - name: controllers    
               range: [1,2]
            - name: morebrokers
               range: [7,8)
```

would flatten to `[1,2,4,5,7]` and therefore the offset for broker 1 is 0 and the offset for broker 4 is 2 etc.

**Comment by k-wall** [comment]
Yes, exactly, flattening the ranges into an array and calculating the index of the node id in that array.   Thank you for the clarification.

**Comment by k-wall** [comment]
Fixed by #1220.  Thank you @robobario
